### ISSUE TITLE
SDL_mixer misdetecting static library as dynamic library during build

### ISSUE THREAD

**Issue by ssokolow** [libsdl-org/SDL_mixer#446]
I'm trying to set up Flatpak packaging for PySolFC, which depends on SDL_mixer via PyGame, and I've run into an odd misbehaviour in the dependency detection:

```
-- Found opusfile: /usr/lib/x86_64-linux-gnu/libopusfile.so  
-- Dynamic opus (opusfile): libopusfile.so.0
-- Enabled ogg music: using stb_vorbis
-- Using system libflac
-- Found FLAC: /usr/lib/x86_64-linux-gnu/libFLAC.so  
-- Dynamic libflac: libFLAC.so.8
-- Using system libmodplug
-- Found modplug: /app/lib/libmodplug.a  
-- Dynamic modplug: libmodplug.a
```

Note those last three lines. They result in this error when running the minimal reproducer build using `playmus.c`:

```
INFO: Opened audio at 44100 Hz 16 bit stereo 4096 bytes audio buffer
INFO: Couldn't load /app/bin/Astral_Dreams.it: Failed loading libmodplug.a: libmodplug.a: cannot open shared object file: No such file or directory
```

...which shows two issues:

1. As was foreshadowed, it's trying to load `libmodplug.a` as if it were a `.so` file.
2. For some reason, it's not picking up the `/app/lib` directory at runtime.

Omitting `*.a` from Flatpak's `cleanup` list and manually setting `LD_LIBRARY_PATH` to make it aware of `/app/lib` results in the following error, as expected:

```
[📦 com.github.libsdl_org.SDL_mixer.PlayMus ~]$ LD_LIBRARY_PATH=/app/lib playmus /app/bin/Astral_Dreams.it 
INFO: Opened audio at 44100 Hz 16 bit stereo 4096 bytes audio buffer
INFO: Couldn't load /app/bin/Astral_Dreams.it: Failed loading libmodplug.a: /app/lib/libmodplug.a: invalid ELF header
```

Now, it's entirely possible that I mis-built libmodplug, given that I just wrote this flatpak-builder source definition to avoid the binary "vendor all/none" option (I'm just trying to replace the SDL_mixer that comes with the Flatpak Runtime with a version that has libmodplug and FluidSynth support enabled, so I'd prefer to depend on the system libraries for everything else) and let it do the rest...

```json
    {
      "name": "libmodplug_sdl_vendored",
      "buildsystem": "cmake",
      "sources": [
        {
          "type": "git",
          "url": "https://github.com/libsdl-org/libmodplug",
          "commit": "f3173c24dcce28d54429099c555665a59d729546"
        }
      ]
    },

```

...but I'd call it a bug that SDL_mixer is picking up the `.a` file as a `.so` file at build time and producing a fundamentally broken runtime artifact.

**Comment by sezero** [comment]
@madebr?

**Comment by ssokolow** [comment]
Here's a minimized version of the `flatpak-builder` manifest I was using:

https://gist.github.com/ssokolow/37b2cb6e7e75aff09711a11ce1557d06

If you've got Flatpak installed and a source with `org.freedesktop.Sdk` configured (eg. Flathub), duplicating my results should be as simple as:

```sh
flatpak install runtime/org.freedesktop.Sdk/x86_64/22.08
flatpak-builder --user --install --force-clean build com.github.libsdl_org.SDL_mixer.PlayMusTest.json
flatpak run com.github.libsdl_org.SDL_mixer.PlayMusTest
```

**Comment by sezero** [comment]
The cmake'ry in libmodplug doesn't default BUILD_SHARED_LIBS to ON.
But that's _no_ excuse that the SDL cmake'ry is finding `*.a` instead of `*.so*`

**Comment by ssokolow** [comment]
Adding a `"config-opts": [ "-DBUILD_SHARED_LIBS=ON" ],` for libmodplug got the correct build-time output...

> -- Using system libmodplug
> -- Found modplug: /app/lib/libmodplug.so  
> -- Dynamic modplug: libmodplug.so.1.0.0

...but I still have to manually add `LD_LIBRARY_PATH=/app/lib` or it won't find it at runtime. 

Is that intended behaviour or should I open another bug?

**Comment by sezero** [comment]
> ...but I still have to manually add `LD_LIBRARY_PATH=/app/lib` or it won't find it at runtime.
> Is that intended behaviour or should I open another bug?

I have no idea about flatpak, so I wouldn't know. Others here would.

We still have to fix the misbehavior of finding a static archive instead of
a dynamic lib.

**Comment by ssokolow** [comment]
I was actually asking about how it found it at compile time but not runtime without `LD_LIBRARY_PATH`, but it now occurs to me that `flatpak-builder` might be setting `LD_LIBRARY_PATH` where `flatpak run` doesn't, so I'll check that.

**Comment by madebr** [comment]
I've added https://github.com/libsdl-org/SDL_mixer/pull/447 that should address this issue.

opusfile, when used as a dynamically loaded library (`dlopen`) is found in 2 stages:

1. find the library using [`find_library`](https://cmake.org/cmake/help/latest/command/find_library.html):
    https://github.com/libsdl-org/SDL_mixer/blob/6cfda9fafb435848010c120da2ba872ba7dccd55/cmake/Findopusfile.cmake#L3-L5
    (This code is shared between linking to opusfile at compile time + dynamically loading it)
2. calculate a [versioned .so name](https://github.com/libsdl-org/SDL_mixer/blob/6cfda9fafb435848010c120da2ba872ba7dccd55/cmake/PrivateSdlFunctions.cmake#L138-L212), which is only needed when dynamically loading it.

What's going wrong in the flatpak build is that cmake is unable to find `libopusfile.so`, which usually has priority over `libopusfile.a`.
#447 will not fix this as cmake won't magically know where to find all libraries.
The actual opsufile library can be overridden by modifying the `opusfile_LIBRARY` cmake variable (by passing `-Dopusfile_LIBRARY=...` during cmake configuration or using ccmake/cmake-gui).

**Comment by ssokolow** [comment]
> https://github.com/libsdl-org/SDL_mixer/pull/447 will not fix this as cmake won't magically know where to find a shared libmodplug.so library.
> The actual modplug library can be overridden by modifying the modplug_LIBRARY cmake variable.
> This can be done by passing `-Dmodplug_LIBRARY=...` during cmake configuration or using ccmake/cmake-gui.

It found it correctly at compile time once I asked libmodplug to build a `.so`, so that side of things is more about making sure that SDL_mixer's cmakery will tell people something's wrong correctly.

> Also noteworthy is to check whether your [CMAKE_SYSTEM_PREFIX_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_PREFIX_PATH.html) is configured correctly.

Probably not relevant here unless its absence/incorrectness is the reason SDL_mixer requires me to manually set `LD_LIBRARY_PATH` in a wrapper for it to find at runtime what it had no problem finding at build time. (Does SDL_mixer's looking for libmodplug have anything RPATH-like that can be enabled?)

**Comment by madebr** [comment]
> Probably not relevant here unless its absence/incorrectness is the reason SDL_mixer requires me to manually set `LD_LIBRARY_PATH` in a wrapper for it to find at runtime what it had no problem finding at build time.
> 
> (Does SDL_mixer's search for libmodplug have anything RPATH-like that can be enabled? Failing that, is there a way to ask for _only_ libmodplug and, if it's used as a library, FluidSynth, to be statically linked, leaving the rest up to reusing the system libraries the Freedesktop Runtime's SDL_mixer was built with support for? I'm a bit unclear on where to look for documentation on this sort of thing.)

Probably a coincidence, but I created #449 asking something similar.
TL;DR is that you need to add the rpath of the dynamically loaded libraries to `libSDL2.so` instead to `libSDL2_mixer.so` because `dlopen` is checking the rpath of "the calling object (i.e., the shared library or executable from which dlopen() is called)" which is `libSDL2.so`.

Of course, you can also just disable dynamical loading (the `SDL2MIXER_*_SHARED` options) and let SDL_mixer link to libmodplug during build time.

**Comment by ssokolow** [comment]
Well, given that `libSDL2.so` is a platform abstraction point that I want to leave the Freedesktop Runtime in charge of updating, I won't be doing the former, but I may experiemnt with doing the latter to replace the launcher script I wrote if I still have the motivation to be a perfectionist after I finish getting PySolFC's solvers building under flatpak-builder.

(Both CMake's documentation and shlomif's custom "my common stuff" CMake module aren't the friendliest to a CMake novice needing to install CMake modules outside `/usr` in a build environment that doesn't allow build-time dependency downloading.)

Thanks for pointing me at the correct options.
